File: dashboard/backend/api/v1/notifications.py
"""Notification API routes for QA-FRAMEWORK Dashboard."""
from fastapi import APIRouter, Depends, HTTPException, Query
from sqlalchemy.ext.asyncio import AsyncSession
from typing import List

from database import get_db
from services.auth_service import get_current_user
from services.notification_service import NotificationService
from models.user import User
from pydantic import BaseModel


router = APIRouter(prefix="/notifications", tags=["notifications"])


class NotificationResponse(BaseModel):
    id: str
    user_id: int
    type: str
    title: str
    message: str
    data: dict
    read: bool
    created_at: str

    class Config:
        from_attributes = True


class NotificationsListResponse(BaseModel):
    notifications: List[NotificationResponse]
    total: int
    unread_count: int


@router.get("", response_model=NotificationsListResponse)
async def get_notifications(
    unread_only: bool = Query(False, description="Filter unread notifications only"),
    limit: int = Query(50, ge=1, le=100, description="Max results"),
    offset: int = Query(0, ge=0, description="Pagination offset"),
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Get all notifications for current user."""
    notifications, total, unread_count = await NotificationService.get_user_notifications(
        db=db,
        user_id=current_user.id,
        unread_only=unread_only,
        limit=limit,
        offset=offset
    )

    return {
        "notifications": [NotificationResponse(**n.to_dict()) for n in notifications],
        "total": total,
        "unread_count": unread_count
    }


@router.post("/{notification_id}/read")
async def mark_notification_as_read(
    notification_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Mark a notification as read."""
    success = await NotificationService.mark_as_read(
        db=db,
        notification_id=notification_id,
        user_id=current_user.id
    )

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"success": True}


@router.post("/read-all")
async def mark_all_notifications_as_read(
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Mark all notifications as read."""
    updated_count = await NotificationService.mark_all_as_read(
        db=db,
        user_id=current_user.id
    )

    return {"updated_count": updated_count}


@router.delete("/{notification_id}")
async def delete_notification(
    notification_id: str,
    current_user: User = Depends(get_current_user),
    db: AsyncSession = Depends(get_db)
):
    """Delete a notification."""
    success = await NotificationService.delete_notification(
        db=db,
        notification_id=notification_id,
        user_id=current_user.id
    )

    if not success:
        raise HTTPException(status_code=404, detail="Notification not found")

    return {"success": True}

File: dashboard/frontend/src/api/notifications.ts
import apiClient from './client'

export interface Notification {
  id: string
  type: 'test_completed' | 'test_failed' | 'suite_created' | 'billing' | 'system'
  title: string
  message: string
  data?: Record<string, any>
  read: boolean
  created_at: string
}

export interface NotificationsResponse {
  notifications: Notification[]
  total: number
  unread_count: number
}

export interface NotificationMarkReadResponse {
  success: boolean
}

export interface NotificationMarkAllReadResponse {
  updated_count: number
}

export const notificationsAPI = {
  /**
   * Get all notifications for current user
   */
  getAll: (params?: {
    unread_only?: boolean
    limit?: number
    offset?: number
  }) => apiClient.get<NotificationsResponse>('/notifications', { params }),

  /**
   * Mark a notification as read
   */
  markAsRead: (id: string) =>
    apiClient.post<NotificationMarkReadResponse>(`/notifications/${id}/read`),

  /**
   * Mark all notifications as read
   */
  markAllAsRead: () =>
    apiClient.post<NotificationMarkAllReadResponse>('/notifications/read-all'),

  /**
   * Delete a notification
   */
  delete: (id: string) =>
    apiClient.delete<{ success: boolean }>(`/notifications/${id}`),
}

export default notificationsAPI
